File: 0048-rotate-image/0048-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    let left = 0
    let right = matrix.length - 1

    while(left < right){
        for(let i = 0; i < right - left; i++){
            let top = left
            let bottom = right
            // save top left
            const topLeft = matrix[top][left + i]

            // move bottom left into top left
            matrix[top][left + i] = matrix[bottom - i][left]

            // move bottom right into bottom left
            matrix[bottom - i][left] = matrix[bottom][right - i]

            // move top right into bottom right
            matrix[bottom][right - i] = matrix[top + i][right]

            //move top left into top right
            matrix[top + i][right] = topLeft
        }
        // move left and right
        left++
        right--
    }
    return matrix
};

File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        top_r = 0
        bottom_r = len(matrix)-1

        while top_r <= bottom_r:
            temp = matrix[top_r]
            matrix[top_r] = matrix[bottom_r]
            matrix[bottom_r] = temp
            top_r += 1
            bottom_r -= 1

        for i in range(len(matrix)):
            for j in range(i, len(matrix[0])):
                temp = matrix[i][j]
                matrix[i][j] = matrix[j][i]
                matrix[j][i] = temp
        
        return matrix